File: backend/auth-service/src/services/auth_service.py
"""
Servicio de Autenticación para Auth Service
Lógica de negocio para registro, login, logout, etc.
Autor: Odalis Senge
"""

from datetime import datetime, timedelta
from typing import Optional, Dict
from src.models.user import User, UserCreate, UserLogin, UserResponse
from src.models.token import RefreshToken, RevokedToken, TokenResponse
from src.services.jwt_service import JWTService
from src.config.database import get_database
from bson import ObjectId


class AuthService:
    """Servicio de autenticación"""
    
    def __init__(self):
        self.db = get_database()
        self.jwt_service = JWTService()
    
    async def register_user(self, user_data: UserCreate) -> UserResponse:
        """Registrar nuevo usuario"""
        # Verificar si el email ya existe
        existing_user = await self.db.users.find_one({"email": user_data.email})
        if existing_user:
            raise ValueError("El email ya está registrado")
        
        # Crear usuario
        hashed_password = self.jwt_service.get_password_hash(user_data.password)
        
        new_user = {
            "email": user_data.email,
            "hashed_password": hashed_password,
            "nombre": user_data.nombre,
            "apellido": user_data.apellido,
            "rol": user_data.rol,
            "activo": True,
            "verificado": False,
            "created_at": datetime.utcnow(),
            "updated_at": datetime.utcnow(),
            "ultimo_acceso": None
        }
        
        result = await self.db.users.insert_one(new_user)
        new_user["_id"] = result.inserted_id
        
        # Retornar respuesta sin password
        return UserResponse(
            id=str(result.inserted_id),
            email=new_user["email"],
            nombre=new_user["nombre"],
            apellido=new_user["apellido"],
            rol=new_user["rol"],
            activo=new_user["activo"],
            verificado=new_user["verificado"],
            created_at=new_user["created_at"],
            ultimo_acceso=new_user["ultimo_acceso"]
        )
    
    async def login_user(self, credentials: UserLogin) -> TokenResponse:
        """Iniciar sesión y generar tokens"""
        # Buscar usuario
        user = await self.db.users.find_one({"email": credentials.email})
        if not user:
            raise ValueError("Credenciales inválidas")
        
        # Verificar contraseña
        if not self.jwt_service.verify_password(credentials.password, user["hashed_password"]):
            raise ValueError("Credenciales inválidas")
        
        # Verificar que el usuario esté activo
        if not user.get("activo", True):
            raise ValueError("Usuario inactivo")
        
        # Generar tokens
        user_id = str(user["_id"])
        token_data = {
            "sub": user_id,
            "email": user["email"],
            "rol": user.get("rol", "user")
        }
        
        access_token = self.jwt_service.create_access_token(token_data)
        refresh_token = self.jwt_service.create_refresh_token({"sub": user_id})
        
        # Guardar refresh token
        refresh_token_doc = {
            "token": refresh_token,
            "user_id": user_id,
            "expires_at": datetime.utcnow() + timedelta(days=7),
            "created_at": datetime.utcnow(),
            "revoked": False
        }
        await self.db.refresh_tokens.insert_one(refresh_token_doc)
        
        # Actualizar último acceso
        await self.db.users.update_one(
            {"_id": user["_id"]},
            {"$set": {"ultimo_acceso": datetime.utcnow()}}
        )
        
        return TokenResponse(
            access_token=access_token,
            refresh_token=refresh_token,
            token_type="Bearer",
            expires_in=15 * 60  # 15 minutos en segundos
        )
    
    async def refresh_access_token(self, refresh_token: str) -> TokenResponse:
        """Renovar access token usando refresh token"""
        # Verificar que el refresh token sea válido
        if not self.jwt_service.verify_token(refresh_token):
            raise ValueError("Refresh token inválido o expirado")
        
        # Verificar que no esté revocado
        revoked = await self.db.revoked_tokens.find_one({"token": refresh_token})
        if revoked:
            raise ValueError("Refresh token revocado")
        
        # Buscar en la BD
        token_doc = await self.db.refresh_tokens.find_one({"token": refresh_token})
        if not token_doc or token_doc.get("revoked"):
            raise ValueError("Refresh token no válido")
        
        # Extraer user_id
        user_id = self.jwt_service.extract_user_id(refresh_token)
        if not user_id:
            raise ValueError("Token inválido")
        
        # Buscar usuario
        user = await self.db.users.find_one({"_id": ObjectId(user_id)})
        if not user:
            raise ValueError("Usuario no encontrado")
        
        # Generar nuevo access token
        token_data = {
            "sub": user_id,
            "email": user["email"],
            "rol": user.get("rol", "user")
        }
        
        new_access_token = self.jwt_service.create_access_token(token_data)
        
        return TokenResponse(
            access_token=new_access_token,
            refresh_token=refresh_token,
            token_type="Bearer",
            expires_in=15 * 60
        )
    
    async def logout_user(self, access_token: str, refresh_token: str) -> bool:
        """Cerrar sesión y revocar tokens"""
        # Extraer user_id del access token
        user_id = self.jwt_service.extract_user_id(access_token)
        if not user_id:
            return False
        
        # Revocar refresh token
        await self.db.refresh_tokens.update_one(
            {"token": refresh_token},
            {"$set": {"revoked": True}}
        )
        
        # Añadir ambos tokens a la blacklist
        access_payload = self.jwt_service.decode_token(access_token)
        refresh_payload = self.jwt_service.decode_token(refresh_token)
        
        revoked_docs = [
            {
                "token": access_token,
                "user_id": user_id,
                "revoked_at": datetime.utcnow(),
                "expires_at": datetime.fromtimestamp(access_payload["exp"]),
                "reason": "logout"
            },
            {
                "token": refresh_token,
                "user_id": user_id,
                "revoked_at": datetime.utcnow(),
                "expires_at": datetime.fromtimestamp(refresh_payload["exp"]),
                "reason": "logout"
            }
        ]
        
        await self.db.revoked_tokens.insert_many(revoked_docs)
        
        return True
    
    async def get_current_user(self, token: str) -> Optional[UserResponse]:
        """Obtener usuario actual desde token"""
        user_id = self.jwt_service.extract_user_id(token)
        if not user_id:
            return None
        
        user = await self.db.users.find_one({"_id": ObjectId(user_id)})
        if not user:
            return None
        
        return UserResponse(
            id=str(user["_id"]),
            email=user["email"],
            nombre=user["nombre"],
            apellido=user.get("apellido"),
            rol=user.get("rol", "user"),
            activo=user.get("activo", True),
            verificado=user.get("verificado", False),
            created_at=user["created_at"],
            ultimo_acceso=user.get("ultimo_acceso")
        )
    
    async def validate_token(self, token: str) -> bool:
        """Validar token (endpoint interno)"""
        # Verificar que el token sea válido
        if not self.jwt_service.verify_token(token):
            return False
        
        # Verificar que no esté en la blacklist
        revoked = await self.db.revoked_tokens.find_one({"token": token})
        if revoked:
            return False
        
        return True

File: backend/auth-service/src/services/auth.service.ts
import jwt from 'jsonwebtoken';
import bcrypt from 'bcrypt';
import { v4 as uuidv4 } from 'uuid';
import { AppDataSource } from '../main';
import { User } from '../entities/user.entity';
import { RefreshToken } from '../entities/refresh-token.entity';
import { TokenBlacklist } from '../entities/token-blacklist.entity';
import {
  jwtConfig,
  bcryptConfig,
  validatePassword,
  validateEmail,
} from '../config/jwt';
import { getRedisClient } from '../config/redis';
import crypto from 'crypto';

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  expiresIn: number;
}

export interface TokenPayload {
  userId: string;
  email: string;
  iat: number;
  exp: number;
}

export class AuthService {
  private userRepository = AppDataSource.getRepository(User);
  private refreshTokenRepository = AppDataSource.getRepository(RefreshToken);
  private blacklistRepository = AppDataSource.getRepository(TokenBlacklist);

  async register(
    email: string,
    password: string,
    name?: string
  ): Promise<{ userId: string; email: string }> {
    // Validar email
    if (!validateEmail(email)) {
      throw new Error('Email inválido');
    }

    // Validar contraseña
    const passwordValidation = validatePassword(password);
    if (!passwordValidation.valid) {
      throw new Error(passwordValidation.error);
    }

    // Verificar si email ya existe
    const existingUser = await this.userRepository.findOne({
      where: { email },
    });
    if (existingUser) {
      throw new Error('El email ya está registrado');
    }

    // Hash de contraseña
    const passwordHash = await bcrypt.hash(password, bcryptConfig.saltRounds);

    // Crear usuario
    const user = this.userRepository.create({
      email,
      passwordHash,
      name: name || email.split('@')[0],
      isActive: true,
    });

    const savedUser = await this.userRepository.save(user);

    console.log(`👤 Usuario registrado: ${savedUser.email}`);

    return {
      userId: savedUser.id,
      email: savedUser.email,
    };
  }

  async login(email: string, password: string): Promise<AuthTokens> {
    // Buscar usuario
    const user = await this.userRepository.findOne({ where: { email } });
    if (!user) {
      throw new Error('Credenciales inválidas');
    }

    if (!user.isActive) {
      throw new Error('Usuario desactivado');
    }

    // Verificar contraseña
    const passwordMatch = await bcrypt.compare(password, user.passwordHash);
    if (!passwordMatch) {
      throw new Error('Credenciales inválidas');
    }

    // Generar tokens
    const tokens = this.generateTokens(user.id, user.email);

    // Guardar refresh token
    const tokenHash = crypto
      .createHash('sha256')
      .update(tokens.refreshToken)
      .digest('hex');
    const expiresAt = new Date(Date.now() + jwtConfig.refreshExpiration * 1000);

    const refreshToken = this.refreshTokenRepository.create({
      userId: user.id,
      tokenHash,
      expiresAt,
    });

    await this.refreshTokenRepository.save(refreshToken);

    // Almacenar en Redis para validación rápida
    const redis = await getRedisClient();
    if (redis) {
      await redis.setEx(
        `rt:${user.id}:${tokenHash}`,
        jwtConfig.refreshExpiration,
        'valid'
      );
    }

    console.log(`🔐 Usuario logueado: ${user.email}`);

    return tokens;
  }

  async refresh(refreshToken: string): Promise<AuthTokens> {
    try {
      // Decodificar sin verificar por ahora
      const decoded = jwt.decode(refreshToken) as TokenPayload;

      if (!decoded || !decoded.userId) {
        throw new Error('Token inválido');
      }

      // Verificar que el token no esté expirado
      if (decoded.exp * 1000 < Date.now()) {
        throw new Error('Refresh token expirado');
      }

      // Buscar usuario
      const user = await this.userRepository.findOne({
        where: { id: decoded.userId },
      });
      if (!user || !user.isActive) {
        throw new Error('Usuario no encontrado o inactivo');
      }

      // Verificar que el refresh token existe en la BD
      const tokenHash = crypto
        .createHash('sha256')
        .update(refreshToken)
        .digest('hex');
      const savedToken = await this.refreshTokenRepository.findOne({
        where: { userId: user.id, tokenHash, revoked: false },
      });

      if (!savedToken) {
        throw new Error('Refresh token revocado o no encontrado');
      }

      // Generar nuevos tokens
      const newTokens = this.generateTokens(user.id, user.email);

      // Guardar nuevo refresh token
      const newTokenHash = crypto
        .createHash('sha256')
        .update(newTokens.refreshToken)
        .digest('hex');
      const newExpiresAt = new Date(
        Date.now() + jwtConfig.refreshExpiration * 1000
      );

      const newRefreshToken = this.refreshTokenRepository.create({
        userId: user.id,
        tokenHash: newTokenHash,
        expiresAt: newExpiresAt,
      });

      await this.refreshTokenRepository.save(newRefreshToken);

      // Revocar token anterior
      savedToken.revoked = true;
      await this.refreshTokenRepository.save(savedToken);

      // Actualizar Redis
      const redis = await getRedisClient();
      if (redis) {
        await redis.del(`rt:${user.id}:${tokenHash}`);
        await redis.setEx(
          `rt:${user.id}:${newTokenHash}`,
          jwtConfig.refreshExpiration,
          'valid'
        );
      }

      console.log(`🔄 Token refrescado para usuario: ${user.email}`);

      return newTokens;
    } catch (error) {
      throw new Error(`Error refrescando token: ${error.message}`);
    }
  }

  async logout(userId: string, accessToken: string): Promise<void> {
    // Revocar todos los refresh tokens del usuario
    const refreshTokens = await this.refreshTokenRepository.find({
      where: { userId, revoked: false },
    });

    for (const token of refreshTokens) {
      token.revoked = true;
      await this.refreshTokenRepository.save(token);
    }

    // Agregar accessToken a blacklist
    try {
      const decoded = jwt.decode(accessToken) as TokenPayload;
      if (decoded && decoded.exp) {
        const tokenHash = crypto
          .createHash('sha256')
          .update(accessToken)
          .digest('hex');
        const expiresAt = new Date(decoded.exp * 1000);

        const blacklistedToken = this.blacklistRepository.create({
          tokenHash,
          expiresAt,
          reason: 'logout',
        });

        await this.blacklistRepository.save(blacklistedToken);

        // Agregar a Redis para validación rápida
        const redis = await getRedisClient();
        if (redis) {
          const ttl = Math.ceil((expiresAt.getTime() - Date.now()) / 1000);
          if (ttl > 0) {
            await redis.setEx(`bl:${tokenHash}`, ttl, 'blacklisted');
          }
        }
      }
    } catch (error) {
      console.error('Error agregando token a blacklist:', error);
    }

    // Limpiar Redis
    const redis = await getRedisClient();
    if (redis) {
      for (const token of refreshTokens) {
        await redis.del(`rt:${userId}:${token.tokenHash}`);
      }
    }

    console.log(`👋 Usuario deslogueado: ${userId}`);
  }

  async getUser(
    userId: string
  ): Promise<{ id: string; email: string; name: string }> {
    const user = await this.userRepository.findOne({ where: { id: userId } });

    if (!user) {
      throw new Error('Usuario no encontrado');
    }

    return {
      id: user.id,
      email: user.email,
      name: user.name,
    };
  }

  async validateToken(token: string): Promise<TokenPayload> {
    try {
      // Verificar en blacklist primero
      const tokenHash = crypto.createHash('sha256').update(token).digest('hex');
      const blacklisted = await this.blacklistRepository.findOne({
        where: { tokenHash },
      });

      if (blacklisted) {
        throw new Error('Token revocado');
      }

      // Verificar JWT
      const decoded = jwt.verify(token, jwtConfig.secret) as TokenPayload;
      return decoded;
    } catch (error) {
      throw new Error(`Token inválido: ${error.message}`);
    }
  }

  private generateTokens(userId: string, email: string): AuthTokens {
    const accessToken = jwt.sign(
      {
        userId,
        email,
      },
      jwtConfig.secret,
      {
        expiresIn: jwtConfig.expiration,
      }
    );

    const refreshToken = jwt.sign(
      {
        userId,
        email,
      },
      jwtConfig.refreshSecret,
      {
        expiresIn: jwtConfig.refreshExpiration,
      }
    );

    return {
      accessToken,
      refreshToken,
      expiresIn: jwtConfig.expiration,
    };
  }
}

export const authService = new AuthService();
